url28dfd00382_byang223_TurnBAPI_tgz-pJ8-units_upgrades_UpgradeTestJ8/src/input/TBInputEngine.java

```diff
@@ -0,0 +1,240 @@
+package input;
+
+import java.awt.event.MouseEvent;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.DataInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileWriter;
+import java.io.InputStreamReader;
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Timer;
+import java.util.TimerTask;
+
+import modes.GameMode;
+import modes.TBGameMode;
+
+import com.golden.gamedev.engine.BaseInput;
+/**
+ * Abstract super class that input engines for
+ * any Turn-Based game should extend and implement
+ * three methods
+ * @author prithvi
+ *
+ */
+public abstract class TBInputEngine {
+	
+	/*
+	 * On instantiation:
+	 * Creates new eventList
+	 * Sets GTGE's BaseInput object as its own
+	 * Loads the arraylist of UserEvents
+	 */
+	public TBInputEngine(BaseInput b, TBGameMode g) {
+		setEventList(new ArrayList<UserEvents>());
+		setBsInput(b);
+		loadEvents(g);
+	}
+
+	/**
+	 * Field for holding list of UserEvents
+	 */
+	private ArrayList<UserEvents> eventList;
+	/**
+	 * Field for GTGE's BaseInput object
+	 */
+	private BaseInput bsInput;
+	
+	/**
+	 * Abstract method to be implemented for a specific game
+	 * to set default controls
+	 * @param map
+	 */
+	public abstract void createDefaultKeyBindingsMap(HashMap<String, Integer> map);
+	
+	/**
+	 * Abstract method determined by specific game
+	 * for deciding on the name of the Key Binding
+	 * file to save to
+	 * @return
+	 */
+	public abstract String getFileName();
+	
+	/**
+	 * Length of time that determines how
+	 * often to check if a file changed
+	 * @return
+	 */
+	public abstract int getFileCheckTime();
+
+	/**
+	 * Create FileWatcher to listen for file changes
+	 * Load key bindings depending on existence of file
+	 * NOTE: If file exists, must exist in correct format
+	 * @param g
+	 */
+	protected void loadEvents(TBGameMode g) {
+		File file=new File(getFileName());
+		TimerTask task = new FileWatcher(file, g);
+		Timer timer = new Timer();
+	    timer.schedule(task , new Date(), getFileCheckTime());
+	    if (!file.exists()) {
+	    	loadDefaultKeyBindings(g, file);
+	    }else{
+	    	loadKeyBindingsFromFile(g, file);
+	    }
+	}
+
+	/**
+	 * Reads from key bindings file and instantiates every class
+	 * with appropriate key
+	 * Assumes file is in correct format (does no error checking)
+	 * @param g
+	 * @param file
+	 * Reference: http://stackoverflow.com/questions/3671649/java-newinstance-of-class-thas-has-no-default-constructor
+	 */
+	protected void loadKeyBindingsFromFile(TBGameMode g, File file) {
+		getEventList().clear();
+		try{
+			FileInputStream fstream = new FileInputStream(file);
+			DataInputStream in = new DataInputStream(fstream);
+			BufferedReader br = new BufferedReader(new InputStreamReader(in));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *br*. The resource is referenced at lines 104, 105, 107. The resource is closed at line 113. There are other execution paths that do not contain closure statements, for example, when Integer.parseInt throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).